src/SharedEventManager.php

```diff
@@ -177,23 +177,47 @@ public function getListeners(array $identifiers, $eventName)
                 ));
             }
 
-            $listenersByIdentifier = isset($this->identifiers[$identifier]) ? $this->identifiers[$identifier] : [];
-
-            $listeners = array_merge_recursive(
-                $listeners,
-                isset($listenersByIdentifier[$eventName]) ? $listenersByIdentifier[$eventName] : [],
-                isset($listenersByIdentifier['*']) ? $listenersByIdentifier['*'] : []
-            );
+            if (isset($this->identifiers[$identifier])) {
+                $listenersByIdentifier = $this->identifiers[$identifier];
+                if (isset($listenersByIdentifier[$eventName])) {
+                    foreach ($listenersByIdentifier[$eventName] as $p => $l) {
+                        $listeners[$p][] = $l;
+                    }
+                }
+                if (isset($listenersByIdentifier['*'])) {
+                    foreach ($listenersByIdentifier['*'] as $p => $l) {
+                        $listeners[$p][] = $l;
+                    }
+                }
+            }
         }
 
-        if (isset($this->identifiers['*']) && ! in_array('*', $identifiers)) {
+        if (isset($this->identifiers['*'])) {
             $wildcardIdentifier = $this->identifiers['*'];
+            if (isset($wildcardIdentifier[$eventName])) {
+                foreach ($wildcardIdentifier[$eventName] as $p => $l) {
+                    $listeners[$p][] = $l;
+                }
+            }
+            if (isset($wildcardIdentifier['*'])) {
+                foreach ($wildcardIdentifier['*'] as $p => $l) {
+                    $listeners[$p][] = $l;
+                }
+            }
+        }
 
-            $listeners = array_merge_recursive(
-                $listeners,
-                isset($wildcardIdentifier[$eventName]) ? $wildcardIdentifier[$eventName] : [],
-                isset($wildcardIdentifier['*']) ? $wildcardIdentifier['*'] : []
-            );
+        if (PHP_VERSION_ID >= 50600) {
+            foreach ($listeners as $p => $l) {
+                $listeners[$p] = array_merge(...$l);
```

Comment:
@marc-mabe I don't think that you will be able to get rid of the 5.6 dependency in an effective way here... I'd totally be for this though (moving forward with the PHP version) if the performance improvement is noticeable